services/storage/src/simcore_service_storage/dsm.py

```diff
@@ -963,10 +979,46 @@ async def search_files_starting_with(
             )
 
             async for row in conn.execute(stmt):
-                meta = FileMetaData(**dict(row))
-                meta_extended = FileMetaDataEx(
-                    fmd=meta,
-                    parent_id=str(Path(meta.object_name).parent),
-                )
+                meta_extended = to_meta_extended(row)
                 files_meta.append(meta_extended)
+
         return list(files_meta)
+
+    async def create_soft_link(
+        self, user_id: int, target_uuid: str, link_uuid: str
+    ) -> FileMetaDataEx:
+
+        # validate link_uuid
+        async with self.engine.acquire() as conn:
+            # TODO: select exists(select 1 from file_metadat where file_uuid=12)
+            found = await conn.scalar(
+                sa.select([file_meta_data.c.file_uuid]).where(
+                    file_meta_data.c.file_uuid == link_uuid
+                )
+            )
+            if found:
+                raise ValueError(f"Invalid link {link_uuid}. Link already exists")
+
+        # validate target_uuid
+        target = await self.list_file(str(user_id), SIMCORE_S3_STR, target_uuid)
+        if not target:
+            raise ValueError(
+                f"Invalid target '{target_uuid}'. File does not exists for this user"
+            )
+
+        # duplicate target and change the following columns:
+        target.fmd.file_uuid = link_uuid
+        target.fmd.file_id = link_uuid  # NOTE: api-server relies on this id
+        target.fmd.is_soft_link = True
+        # target.fmd.file_id = link_uuid
+
+        async with self.engine.acquire() as conn:
+            stmt = (
+                file_meta_data.insert()
+                .values(**attr.asdict(target.fmd))
+                .returning(literal_column("*"))
+            )
```

Comment:
why not use an upset, in case of concurrency issues?

Reply:
I tried to emulate the ``ln`` tool in linux. If you try to rename an existing link it fails ...
```cmd
$ ln -s README.md README-link.md
$ ln -s docker-compose.yml README-link.md 
ln: failed to create symbolic link 'README-link.md': File exists
```
I do the same here. 

In any case, i agree that the error handling is far from ideal. I refer again to the coming PRs to do that separately for clarity.